dashboard/app/models/contact_rollups_processed.rb

```diff
@@ -0,0 +1,94 @@
+  #
+  # @param [Integer] batch_size number of records to save per INSERT statement.
+  def self.import_from_raw_table(batch_size = DEFAULT_BATCH_SIZE)
+    # Combines data and metadata for each record in contact_rollups_raw table into one JSON field.
+    # The query result has the same number of rows as in contact_rollups_raw.
+    select_query = <<-SQL.squish
+      SELECT email, JSON_OBJECT('sources', sources, 'data', data, 'data_updated_at', data_updated_at) AS data_and_metadata
+      FROM contact_rollups_raw
+    SQL
+
+    # Groups records by emails. Aggregates all data and metadata belong to an email into one JSON field.
+    #
+    # Note: use GROUP_CONCAT instead of JSON_OBJECT_AGG because the current Aurora Mysql version in
+    # production is 5.7.12, while JSON_OBJECT_AGG is only available from 5.7.22.
+    # Because GROUP_CONCAT returns a string, we add a parser function to convert the result to a hash.
+    group_by_query = <<-SQL.squish
+      SELECT email, CONCAT('[', GROUP_CONCAT(data_and_metadata), ']') AS all_data_and_metadata
+      FROM (#{select_query}) AS sub_query
+      GROUP BY email
+    SQL
+
+    # Process the aggregated data row by row and save the results to DB.
+    batch = []
+    ActiveRecord::Base.connection.exec_query(group_by_query).each do |contact|
+      contact_data = parse_contact_data(contact['all_data_and_metadata'])
+
+      processed_contact_data = {}
+      processed_contact_data.merge!(extract_opt_in(contact_data) || {})
+
+      batch << {email: contact['email'], data: processed_contact_data}
+
+      if batch.size == batch_size
+        import! batch, validate: false
+        batch = []
+      end
+    end
+
+    import! batch, validate: false unless batch.empty?
+  end
+
+  # Parses a JSON string contains all data and metadata of a contact.
+  # It will throw exception if cannot parse the entire input string.
+  #
+  # @param [String] str represents a JSON array. Each array item is a hash {sources:String, data:Hash, data_updated_at:DateTime}.
+  #
+  # @return [Hash] a hash {source_table => {contact_attribute => value}}
+  def self.parse_contact_data(str)
+    parsed_items = JSON.parse(str)
+
+    {}.tap do |output|
+      parsed_items.each do |item|
+        # In a valid item, only data value could be null
+        sources = item['sources']
+        data = item['data'] || {}
+        data_updated_at = Time.parse(item['data_updated_at'])
+
+        output[sources] = data.merge('data_updated_at' => data_updated_at)
+      end
+    end
+  end
+
+  # Extracts opt_in info from contact data.
+  #
+  # @param [Hash] contact_data compiled data from multiple source tables.
+  #   Input hash structure: {source_table => {contact_attribute => value}}
+  #
+  # @return [Hash, nil] a hash containing opt_in key and value (could be nil)
+  #   or nil if opt_in does not exist in the input.
+  def self.extract_opt_in(contact_data)
```

Comment:
Sorry I think you explained this but I forget, will there be separate extract functions per source? Or will this be generalized in the future?

Reply:
There will be one extract function per contact attribute, e.g., `extract_state` and `extract_country`. Those functions will share some common steps such as checking if an attribute exist or not, which then be generalized.

Reply:
This is a half baked thought for future consideration when we have multiple sources but maybe could be done with a hash that lists for each source the expected keys, and a single function that does this step?

Reply:
I'm not quite clear how that will look like yet. We can wait until we have more extraction functions to decide how to optimize them (for performance or for readability). There is a quote from DHH (Rails creator) that can be applicable here _"The best frameworks are in my opinion extracted, not envisioned. And the best way to extract is first to actually do."_ ([link](https://dhh.dk/posts/6-why-theres-no-rails-inc)).